AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSCoreDataClient.m

```diff
@@ -0,0 +1,235 @@
+                                                                     errorMessage:message
+                                                                       documentId:documentId]);
+          return;
+        }
+
+        // Create a deleted document record.
+        MSDocumentWrapper *createdOrUpdatedDocument = [[MSDocumentWrapper alloc] initWithDeserializedValue:document
+                                                                                                 jsonValue:jsonDocument
+                                                                                                 partition:token.partition
+                                                                                                documentId:documentId
+                                                                                                      eTag:cachedDocument.eTag
+                                                                                           lastUpdatedDate:cachedDocument.lastUpdatedDate
+                                                                                          pendingOperation:operation
+                                                                                                     error:nil];
+
+        // Update local store and return document.
+        [self updateLocalStore:token
+            currentCachedDocument:cachedDocument
+                newCachedDocument:createdOrUpdatedDocument
+                 deviceTimeToLive:deviceTimeToLive
+                        operation:operation];
+        completionHandler(createdOrUpdatedDocument);
+      }
+    }
+  });
+}
+
+/**
+ * Validate an operation.
+ *
+ *@param operation The operation.
+ *
+ *@return YES if the operation is valid; NO otherwise.
+ */
++ (BOOL)isValidOperation:(NSString *)operation {
+  return operation == nil || [kMSPendingOperationCreate isEqualToString:operation] ||
+         [kMSPendingOperationReplace isEqualToString:operation] || [kMSPendingOperationDelete isEqualToString:operation];
+}
+
+/**
+ * Returns a flag indicating if a remote operation should be attempted.
+ *
+ * @param cachedDocument The current cached document (if any).
+ *
+ * @return YES if a remote operation should be attempted; NO otherwise.
+ */
+- (BOOL)needsRemoteOperation:(MSDocumentWrapper *)cachedDocument {
+  return // Not offline
+      [self.reachability currentReachabilityStatus] != NotReachable
+      // And no pending operation on cached document
+      && cachedDocument.pendingOperation == nil;
+}
+
+- (void)updateLocalStore:(MSTokenResult *)token
+    currentCachedDocument:(MSDocumentWrapper *)currentCachedDocument
+        newCachedDocument:(MSDocumentWrapper *)newCachedDocument
+         deviceTimeToLive:(NSInteger)deviceTimeToLive
+                operation:(NSString *_Nullable)operation {
+
+  // If the device time to live does not allow it, do not touch the local storage.
+  if (deviceTimeToLive == MSDataStoreTimeToLiveNoCache) {
```

Comment:
This is not correct. The entry must be deleted from the cache in this case, because now the cache expiration time is technically supposed to become 0.